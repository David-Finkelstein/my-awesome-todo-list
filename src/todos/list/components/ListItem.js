import React from 'react'
import cn from 'classnames';
import PropTypes from 'prop-types';
import { Button } from '@guestyci/atomic-design/dist/components';
import { Link } from 'react-router-dom'


class TodoItem extends React.Component {
    constructor(props){
        super(props);

        this.onTodoDelete = this.onTodoDelete.bind(this);
        this.onToggleTodoStatus = this.onToggleTodoStatus.bind(this);
    }

    onTodoDelete() {
        const { todoProperties: { id }, deleteTodo} = this.props;
        deleteTodo(id);
    }

    onToggleTodoStatus() {
        const { todoProperties: { id }, toggleTodoStatus} = this.props;
        toggleTodoStatus(id);
    }

    render(){
        const { todoProperties: { id ,finished, text }} = this.props;

        return (
            <li className="ui-state-default bg-white border-0 border-bottom-1 pt-3 pr-0" key={id}>
                <div className={cn({"checked": finished})}>
                    <label onClick={this.onToggleTodoStatus}>{text}</label>
                    <Link to={`/todos/${id}`}>
                        <Button className="btn btn-sm ml-5" color="success">Edit</Button>
                    </Link>
                    <Button className="btn btn-sm ml-5" onClick={this.onTodoDelete} color="danger">Delete</Button>
                </div>
            </li>
        );
    }
}

TodoItem.propTypes = {
    todoProperties: PropTypes.shape({
        text: PropTypes.string.isRequired,
        finished: PropTypes.bool.isRequired,
        id: PropTypes.number.isRequired
    }).isRequired,
    deleteTodo: PropTypes.func.isRequired,
    toggleTodoStatus: PropTypes.func.isRequired,
};

export default TodoItem;
