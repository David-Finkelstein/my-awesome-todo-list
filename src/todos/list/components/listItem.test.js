import React from 'react';
import { mount, shallow } from "enzyme";
import { mountToJson } from "enzyme-to-json";
import { BrowserRouter as Router, Route } from "react-router-dom";

import ListItem from './ListItem'

const mockProps = {
    deleteTodo: jest.fn(),
    toggleTodoStatus: jest.fn(),
    todoProperties: {
        text: 'Clean house',
        finished: true,
        id: 12464535
    }
};

describe('<ListItem />', () => {
    it('checks rendering with new props', () => {
        const newTodoProperties = {
            text: 'Clean house',
            finished: true,
            id: 12464535
        };
        const wrapper = mount(
            <Router>
                <Route render={() => <ListItem {...mockProps}/>}/>
            </Router>
        );

        expect(wrapper.find('label')).toBeDefined();
        expect(wrapper.find('label').text()).toBe(mockProps.todoProperties.text);

        wrapper.setProps({todoProperties: newTodoProperties});
        expect(wrapper.find('label').text()).toBe(newTodoProperties.text);
    });

    it('checks the functionality of delete item', () => {
        const wrapper = shallow(<ListItem {...mockProps}/>);
        const deleteBtn = wrapper.find('.btn').at(1);

        expect(mockProps.deleteTodo.mock.calls.length).toBe(0);
        deleteBtn.props().onClick();
        expect(mockProps.deleteTodo.mock.calls.length).toBe(1);
    });

    it('checks the functionality of toggle', () => {
        const wrapper = shallow(<ListItem {...mockProps}/>);
        const toogleText = wrapper.find('label');

        expect(mockProps.toggleTodoStatus.mock.calls.length).toBe(0);
        toogleText.props().onClick();
        expect(mockProps.toggleTodoStatus.mock.calls.length).toBe(1);
    });

    it('takes a snapshot of the component and check if the prev snapshot equals to the current', () => {
        const wrapper = shallow(<ListItem {...mockProps}/>);
        expect(mountToJson(wrapper)).toMatchSnapshot();
    });
});
