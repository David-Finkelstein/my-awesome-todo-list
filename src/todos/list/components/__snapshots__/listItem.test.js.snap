// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem /> takes a snapshot of the component and check if the prev snapshot equals to the current 1`] = `
<li
  className="ui-state-default bg-white border-0 border-bottom-1 pt-3 pr-0"
  key="12464535"
>
  <div
    className="checked"
  >
    <label
      onClick={[Function]}
    >
      Clean house
    </label>
    <Link
      replace={false}
      to="/todos/12464535"
    >
      <Jt
        className="btn btn-sm ml-5"
        color="success"
      >
        Edit
      </Jt>
    </Link>
    <Jt
      className="btn btn-sm ml-5"
      color="danger"
      onClick={[Function]}
    >
      Delete
    </Jt>
  </div>
</li>
`;
